src/main/java/com/microsoft/sqlserver/jdbc/SSPIAuthentication.java

```diff
@@ -149,23 +149,34 @@ String enrichSpnWithRealm(SQLServerConnection con, String spn, boolean allowHost
 
         String dnsName = m.group(1);
         String portOrInstance = m.group(2);
-        if (null == realm || realm.trim().isEmpty()) {
-            RealmValidator realmValidator = getRealmValidator();
-            realm = findRealmFromHostname(realmValidator, dnsName);
-            if (null == realm && allowHostnameCanonicalization) {
-                // We failed, try with canonical host name to find a better match
-                try {
+        try {
+            if (null == realm || realm.trim().isEmpty()) {
+                RealmValidator realmValidator = getRealmValidator();
+                realm = findRealmFromHostname(realmValidator, dnsName);
+                if (null == realm && allowHostnameCanonicalization) {
+                    if (logger.isLoggable(Level.FINER)) {
+                        logger.finer("Realm not provided or found from host name, attempt to find canonical host name via InetAddress function and then derive realm");
+                    }
+                    // We failed, try with canonical host name to find a better match
                     String canonicalHostName = InetAddress.getByName(dnsName).getCanonicalHostName();
                     realm = findRealmFromHostname(realmValidator, canonicalHostName);
-                    // match means hostname is correct (eg if server name was an IP) so override dnsName as well
+                    // Match means hostname is correct (eg if server name was an IP) so override dnsName as well
                     dnsName = canonicalHostName;
-                } catch (UnknownHostException e) {
-                    // ignored, cannot canonicalize
+                }
+            } else {
+                if (allowHostnameCanonicalization) {
                     if (logger.isLoggable(Level.FINER)) {
-                        logger.finer("Could not canonicalize host name. " + e.toString());
+                        logger.finer("Since realm is provided, attempt to get canonical host name with InetAddress function");

```

Comment:
log message here seems a bit misleading, what we are trying to do as per comment below is to resolve cname to hostname